### PR TITLE
Add proptest tests and test utilities module

### PR DESCRIPTION
**Pull request by Nemo157**
## Description
Adds automatically generated tests based on [`proptest`](https://altsysrq.github.io/proptest-book/intro.html) to ensure arbitrary compression works. Also refactors a lot of the existing tests to use common utilities to simplify them.

I'll probably move `PendRead` and `PendStream` upstream to `futures-test` since they're likely useful in other testing, so they can be pulled directly from there on the next release. (https://github.com/rust-lang-nursery/futures-rs/pull/1576)

## Motivation and Context

More thorough testing.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the [CONTRIBUTING](https://github.com/rust-net-web/tide/blob/master/.github/CONTRIBUTING.md) document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: tests/utils/mod.rs [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+    }
+}
+
+impl<R: AsyncRead> AsyncRead for PendRead<R> {
+    fn poll_read(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        buf: &mut [u8],
+    ) -> Poll<io::Result<usize>> {
+        let this = self.project();
```
Comment (pickfire) [inline tests/utils/mod.rs]: I think we should get a variable that is self explanatory compared for `this`.

Comment (Nemo157) [reply tests/utils/mod.rs]: This should really be `let self = self.project();`, but you can't use `self` as a variable name, so `this` is just a synonym for it.

**Comment by Nemo157** [comment]
Now that CI is running and decompression is all tested this should be good to merge, r? @fairingrey

**Comment by fairingrey** [comment]
Looks good to me. I like how much cleaner the tests look to write. Feel free to merge it whenever @Nemo157

**Pull request merged**